src/Riok.Mapperly/Descriptors/Enumerables/EnsureCapacity/EnsureCapacityBuilder.cs

```diff
@@ -0,0 +1,69 @@
+﻿using System.Diagnostics.CodeAnalysis;
+using Microsoft.CodeAnalysis;
+using Riok.Mapperly.Helpers;
+
+namespace Riok.Mapperly.Descriptors.Enumerables.EnsureCapacity;
+
+public static class EnsureCapacityBuilder
+{
+    private const string EnsureCapacityName = "EnsureCapacity";
+    private const string CountPropertyName = nameof(ICollection<object>.Count);
+    private const string LengthPropertyName = nameof(Array.Length);
+    private const string TryGetNonEnumeratedCountMethodName = "TryGetNonEnumeratedCount";
+
+    public static EnsureCapacity? TryBuildEnsureCapacity(ITypeSymbol sourceType, ITypeSymbol targetType, WellKnownTypes types)
+    {
+        var capacityMethod = targetType.GetAllMembers(EnsureCapacityName)
+                        .OfType<IMethodSymbol>()
+                        .FirstOrDefault(x => x.Parameters.Length == 1
+                        && x.Parameters[0].Type.SpecialType == SpecialType.System_Int32
+                        && !x.IsStatic);
+
+        // if EnsureCapacity is not available then return null
+        if (capacityMethod == null)
+            return null;
+
+        // if target does not have a count then return null
+        if (!TryGetNonEnumeratedCount(targetType, types, out var targetSizeProperty))
+            return null;
+
+        // if target and source count are known then create a simple EnsureCapacity statement
+        if (TryGetNonEnumeratedCount(sourceType, types, out var sourceSizeProperty))
+            return new EnsureCapacityMember(targetSizeProperty, sourceSizeProperty);
+
+        if (!sourceType.ImplementsGeneric(types.IEnumerableT, out var iEnumerable))
+            return null;
```

Comment:
Is this ever reached? `TryBuildEnsureCapacity` is only called for enumerables, isn't it?

Reply:
I used ImplementsGeneric to get the typed version of `IEnumerableT`. I originally didn't use this redundant null, but changed it when you [suggested I remove the null suppression](https://github.com/riok/mapperly/pull/312#discussion_r1167799422)